aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java

```diff
@@ -0,0 +1,157 @@
+    private final SQLCommandFactory commandFactory;
+    private final SQLiteDatabase database;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param commandFactory SQL command factory
+     * @param database SQLite database connection handle
+     */
+    SQLiteModelTree(ModelSchemaRegistry registry,
+                    SQLCommandFactory commandFactory,
+                    SQLiteDatabase database) {
+        this.registry = registry;
+        this.commandFactory = commandFactory;
+        this.database = database;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
+        if (Empty.check(root)) {
+            throw new IllegalArgumentException("Cannot traverse tree from an empty root.");
+        }
+        Map<ModelSchema, Set<String>> descendants = new LinkedHashMap<>();
+        ModelSchema rootSchema = registry.getModelSchemaForModelInstance(root.iterator().next());
+        Set<String> rootIds = new HashSet<>();
+        for (T model : root) {
+            rootIds.add(model.getId());
+        }
+        recurseTree(descendants, rootSchema, rootIds);
+        return descendants;
+    }
+
+    private void recurseTree(
+            Map<ModelSchema, Set<String>> tree,
+            ModelSchema modelSchema,
+            Collection<String> parentIds
+    ) {
+        SQLiteTable parentTable = SQLiteTable.fromSchema(modelSchema);
+        for (ModelAssociation association : modelSchema.getAssociations().values()) {
+            switch (association.getName()) {
+                case "HasOne":
+                case "HasMany":
+                    String childModel = association.getAssociatedType(); // model name
+                    ModelSchema childSchema = registry.getModelSchemaForModelClass(childModel);
+                    SQLiteTable childTable = SQLiteTable.fromSchema(childSchema);
+                    String childPrimaryKey = childTable.getPrimaryKey().getAliasedName();
+                    QueryField queryField = QueryField.field(parentTable.getPrimaryKeyColumnName());
+
+                    // Chain predicates with OR operator.
+                    // No predicate = Match NONE.
+                    // 1 predicate = Match SELF.
+                    // 2 or more predicates = Match ANY.
+                    QueryPredicate predicate = QueryPredicates.none();
+                    for (String parentId : parentIds) {
+                        QueryPredicateOperation<Object> operation = queryField.eq(parentId);
```

Comment:
```suggestion
                        QueryPredicateOperation<String> operation = queryField.eq(parentId);
```

This should be the data type of the field being evaluates, which is `String` for the model id.

Reply:
I want to put `String` there as well, but because `QueryField#eq(Object)` was originally declared that way, we can't do that. I considered refactoring `QueryField` class to use generics instead of `Object`, but that is potentially a breaking change, and also outside of this PR's scope.

Reply:
Got it, makes sense!  I like the idea of refactoring `QueryField` to use generics as well.  There was a customer recently who tried to pass a `String` where they should have passed a `Temporal.DateTime` object.  If `QueryField` was generic, it would help customers avoid errors like that.  Like you said, it would be a breaking change though.